rustls/src/crypto/mod.rs

```diff
@@ -505,6 +505,81 @@ pub trait ActiveKeyExchange: Send + Sync {
+    ///
+    /// "Hybrid" means a key exchange algorithm which is constructed from two
+    /// (or more) underlying independent algorithms. Usually one is post-quantum-secure,
+    /// and the other is "classical".  See
+    /// <https://datatracker.ietf.org/doc/draft-ietf-tls-hybrid-design/10/>
+    ///
+    /// There is no requirement for a hybrid scheme (or any other!) to implement
+    /// `hybrid_pair`. It only enables an optimization.
+    ///
+    /// # Background
+    /// Rustls always sends a presumptive key share in its ClientHello, using
+    /// (absent any other information) the first item in [`CryptoProvider::kx_groups`].
+    /// If the server accepts the client's selection, it can complete the handshake
+    /// using that key share.  If not, the server sends a `HelloRetryRequest` instructing
+    /// the client to send a different key share instead.
+    ///
+    /// This costs an extra round trip, and wastes the key exchange computation
+    /// (in [`SupportedKxGroup::start()`]) the client already did.  We would
+    /// like to avoid those wastes if possible.
+    ///
+    /// It is early days for post-quantum-secure hybrid key exchange deployment.
+    /// This means (commonly) continuing to offer both the hybrid and classical
+    /// key exchanges, so the handshake can be completed without a `HelloRetryRequest`
+    /// for servers that support the offered hybrid or classical schemes.
+    ///
+    /// Implementing `hybrid_pair` enables two optimizations:
+    ///
+    /// 1. Sending both the hybrid and classical key shares in the `ClientHello`.
+    ///
+    /// 2. Performing the classical key exchange setup only once.  This is important
+    ///    because the classical key exchange setup is relatively expensive.
+    ///    This optimization is permitted and described in
+    ///    <https://www.ietf.org/archive/id/draft-ietf-tls-hybrid-design-10.html#section-3.2>
+    ///
+    /// Both of these only happen if the classical algorithm appears separately in
+    /// the client's [`CryptoProvider::kx_groups`], and if the hybrid algorithm appears
+    /// first in that list.
+    ///
+    /// # How it works
+    /// This function is only called by rustls for clients.  It is called when
+    /// constructing the initial `ClientHello`.  rustls follows these steps:
+    ///
+    /// 1. If the return value is None, nothing further happens.
+    /// 2. If the given [`ActiveKeyExchange::group()`] does not appear in
+    ///    [`CryptoProvider::kx_groups`], nothing further happens.
+    /// 3. The given [`ActiveKeyExchange::pub_key()`] is added to the `ClientHello`,
+    ///    after the hybrid entry.
+    ///
+    /// Then, one of three things may happen when the server replies to the `ClientHello`:
+    ///
+    /// 1. The server sends a `HelloRetryRequest`.  Everything is thrown away and
+    ///    we start again.  Note that `hybrid_pair` is never called in this case.
```

Comment:
Maybe worth clarifying here. It's been called for the initial handshake and I think the intent is to communicate it won't be called again for the HRR right?

> Note that `hybrid_pair()` is not called again in this case.

I also think we like having parens on fn names.

Reply:
Actually I meant to talk about `into_hybrid_pair()` in this clause.